sshd-core/src/main/java/org/apache/sshd/client/session/ClientSessionImpl.java

```diff
@@ -90,16 +88,17 @@ public ClientSessionImpl(ClientFactoryManager client, IoSession ioSession) throw
         authFuture.setAuthed(false);
 
         signalSessionCreated(ioSession);
-        sendClientIdentification();
 
-        KexExtensionHandler extHandler = getKexExtensionHandler();
-        if ((extHandler == null) || (!extHandler.isKexExtensionsAvailable(this, AvailabilityPhase.PREKEX))) {
-            kexState.set(KexState.INIT);
-            sendKexInit();
-        } else {
-            if (log.isDebugEnabled()) {
-                log.debug("<init>({}) delay KEX-INIT until server-side one received", this);
-            }
+        /*
+         * Must be called regardless of whether the client identification
+         * is sent or not immediately in order to allow opening any underlying
+         * proxy protocol - e.g., SOCKS or HTTP CONNECT - otherwise the server's
+         * identification will never arrive
+         */
+        initializeProxyConnector();
+
+        if (sendImmediateIdentification) {
```

Comment:
A tiny bit ugly that `if (sendImmediateIdentification)` is here in ClientSessionImpl, but the corresponding `if (!sendImmediateIdentification)` is in the super class AbstractClientSession.

Reply:
Agreed, but I did not want to make too many changes...